ktor-client/ktor-client-core/common/src/io/ktor/client/request/HttpRequest.kt

```diff
@@ -117,9 +117,15 @@ class HttpRequestBuilder : HttpMessageBuilder {
      * Mutates [this] copying all the data from another [builder] using it as base.
      */
     @InternalAPI
-    @Deprecated("For internal usage only")
-    fun takeFrom(builder: HttpRequestBuilder): HttpRequestBuilder {
+    fun takeFromWithExecutionContext(builder: HttpRequestBuilder): HttpRequestBuilder {
         executionContext = builder.executionContext
+        return takeFrom(builder)
+    }
+
+    /**
+     * Mutates [this] copying all the data but execution context from another [builder] using it as base.
+     */
+    fun takeFrom(builder: HttpRequestBuilder): HttpRequestBuilder {
```

Comment:
@e5l The reason for the problem is a mistake we did. I believe it is a broken design to keep a job in the builder. It looks even more strange if we notice that the builder could be built multiple times. Can't we eliminate the context (job) at all in the builder? Can't we keep it somewhere else?

Reply:
As far as I see it we use `HttpRequestBuilder` as a context of the request and send pipelines. We pass through all steps and uses them in many places and we use `executionContext` it contains also.

Reply:
I believe that request building and request execution are separate things and only the second is tracked and executed under a job while the first could be prepared at any time or, in theory, even serialized/deserialized.

Reply:
What I am saying is that just a mental model. Currently, we have them mixed up.